### PR TITLE
ISO20022 Ecosystem Research and Technical Design

### PR DESCRIPTION
**Pull request by pifragile**
ISO20022 is the future standard of financial transactions of all kinds including payments, foreign exchange and securities trading. Many blockchain ecosystems like [Ripple](https://ripple.com/), [Cardano](https://cardano.org/) and [Stellar](https://stellar.org/) are already ISO20022 compliant. In order to make Polkadot ISO20022 compatible as well, it is necessary to start designing how such an integration would look like in practice.

The goal of this project is to get an overview of how other blockchain ecosystems approach the ISO20022 integration, outline challenges and opportunities of such an integration and provide technical designs of how Polkadot could become ISO20022 compatible.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [x] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ Polkadot (USDC & USDT) address in the application).
- [x] I am aware that, in order to receive a grant, I (and the entity I represent) have to successfully complete a KYC/KYB check.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Comment by wstrametz** [comment]
Hello from element36 - we support ISO 20022 pain.001 and camt files processed via Ebics protocoll using an Off-chain-worker; our current implementation adds Zero-Knowledge proofs (arguments of knowledge as STARKs to be more precise) for downloaded camt53 files, so independenlty of Ebics. Thus we are able to prove settlement of payments and assets (in/out), or identities in a non-interactive and trustless manner. Happy to jump on a call to carve out differences or synergies, just pm me. Thx, Walter

**Comment by pifragile** [comment]
Thanks for the heads up @keeganquigley, I was not aware of this project yet.
I took a look at their proposal and whitepaper and I can summarize as follows (@wstrametz  correct me if I'm wrong):

Hyperfridge is about making information of banking data available on-chain in a secure and private way and having on-chain events trigger bank transactions. Payments are settled off-chain. As they write in their whitepaper: "It is about triggering and mirroring state changes of a bank account on a ledger".

Our project is different. We want to use the Polkadot blockchain as the actual settlement layer for payments. The benefits thereof are:

1. Bank users can hold and transfer DOT via their banks using ISO20022 messages. Also payment reversals (ie. via. pain.007.001.11 CustomerPaymentReversalV11) will be supported.
2. Banks can use DOT as an intermediate currency for inter-bank payments and thus reduce friction in (cross-border/foreign exchange) payments.

@wstrametz Thank you, sent you a message on Linkedin to schedule a call!

**Comment by wstrametz** [comment]
> Thanks for the heads up @keeganquigley, I was not aware of this project yet. I took a look at their proposal and whitepaper and I can summarize as follows (@wstrametz correct me if I'm wrong):
> 
> Hyperfridge is about making information of banking data available on-chain in a secure and private way and having on-chain events trigger bank transactions. Payments are settled off-chain. As they write in their whitepaper: "It is about triggering and mirroring state changes of a bank account on a ledger".
> 
> Our project is different. We want to use the Polkadot blockchain as the actual settlement layer for payments. The benefits thereof are:
> 
> 1. Bank users can hold and transfer DOT via their banks using ISO20022 messages. Also payment reversals (ie. via. pain.007.001.11 CustomerPaymentReversalV11) will be supported.
> 2. Banks can use DOT as an intermediate currency for inter-bank payments and thus reduce friction in (cross-border/foreign exchange) payments.
> 
> @wstrametz Thank you, sent you a message on Linkedin to schedule a call!

Yes indeed - this is a big difference; hyperfridge is basically a bidirectional trustless bridge to real-world-assets like FIAT sitting on banking ledgers. Your idea I guess is inter-bank communication and settlement similar like R3 consortium ideas or as JP Morgan did with Quorum (now Consensys) as alternative to SWIFT. The biggest challenge for adoption I guess is not techniclal but getting (many) banks on a shared ledger and protocol. Also you need to consider that banks report (sync) their own ledgers to national banks for international transfers, which do the clearing with the "other" national banks - thus they probably need to support DOT as well.

**Comment by pifragile** [comment]
@wstrametz Thanks for your thoughts and inputs!

**Comment by wstrametz** [comment]
> @wstrametz Thanks for your thoughts and inputs!

happy to jump on a call anyway: https://www.linkedin.com/in/strametz/

 A note: the whitepaper is not accurate any more, I learned quite a bit while implementing it. Its simpler and much more generic E.g. no merkle tree needed any more, "composition" (Risc0 feature) will do the trick, transactions and previous states can be chained without keeping a state. Not started yet, but I will create a new version which includes learnings.

**Comment by pifragile** [comment]
> happy to jump on a call anyway: https://www.linkedin.com/in/strametz/

Sent you a contact request.

**Comment by keeganquigley** [comment]
Sorry for the delay @pifragile I marked the application as ready for review and pinged the committee.

**Comment by pifragile** [comment]
Thank you @keeganquigley. I updated the description to include the comparison with Hyperfridge.

File: applications/ISO20022-Implementation-POC.md [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+2. We do not use offchain-workers for message processing also for privacy reasons.
+3. We do not create customer accounts on chain, because they could easily be linked to a real person, again breaking privacy.
+4. We do not use one centralized message processing institution, as we want to give each bank full access to their on-chain funds, at least guaranteeing decentralization on the interbank level.
+
+
+
+
+### Ecosystem Fit
+
+Many competing blockchain ecosystems are already ISO20022 compliant (see above). If Polkadot as well wants to become compatible with the traditional finance industry, an ISO20022 integration will be the right place to start.
```
Comment (semuelle) [inline applications/ISO20022-Implementation-POC.md:70]: Can you give some examples which blockchains are ISO20022 compatible and how they achieved that?

**Review by semuelle — COMMENTED** [review]
Sorry for the long radio silence, @pifragile. My main question would be what your incentive is to build this. Are you planning to use this client yourself? In the `Future Plans` section, you write that this project "can be extended", but not whether you are actually planning to do any of this.

File: applications/ISO20022-Implementation-POC.md (hunk 1 above)
Comment (pifragile) [reply applications/ISO20022-Implementation-POC.md:70]: Blockchains that claim to be ISO20022 compliant are for example Ripple and Cardano. How they actually achieve this is pretty vague in my eyes. For example Ripple writes in their [ISO20022 one pager](https://ripple.com/lp/iso-overview/):

"To help enable this next step in global interoperability and meet the evolving needs of our customers, Ripple is now part of the ISO 20022 Standards Body-the first member focused on Distributed Ledger Technology (DLT)."

What this means on a technical level does not seem very transparent. The way I see it, it is not even totally clear what it means for a blockchain to be ISO20022 compliant, as ISO20022 is a messaging protocol intended to simplify and standardize communication about financial transactions  between trusted entities.

So, I see this proposal as an experiment to bring more clarity to this question by building an ISO20022 integration that is simple and sound and can be used and tested in practice.

**Comment by pifragile** [comment]
> Sorry for the long radio silence, @pifragile. My main question would be what your incentive is to build this. Are you planning to use this client yourself? In the `Future Plans` section, you write that this project "can be extended", but not whether you are actually planning to do any of this.

No worries @[semuelle](https://github.com/semuelle)! My incentive is to find opportunities in the Polkadot ecosystem. I know my way around Substrate and like to design protocols and build stuff. So I wanted to create a grant application for a project that is actually needed in the ecosystem and thus I replied to an RFP. I am not planning to use this client myself.

File: applications/ISO20022-Implementation-POC.md [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+4. Send extrinsic to pallet for reversal
+5. Send/Read ISO messages to bus
+6. Verify relevant ISO messages on chain
+7. Client account management system
+    - Create for each client an account number
+    - Store for each client a list of hashes form the chain
+    - Functionality to derive client balance from onchain hashes
+8. Hashmap-like storage of all ISO messages and salts
+
+We will use NodeJs to build this MVP client.
```
Comment (takahser) [inline applications/ISO20022-Implementation-POC.md:45]: Would you be open to a Java or C#-implementation? Personally, I'd prefer to develop on Node as as well, but I know that banks (at least in Switzerland) are usually quite hesitant to adopt Node and their devs mostly deal with C# and Java. Also, they tend to be more stable when compared to Node w.r.t. potential dependencies, although that can be mitigated by not using any npm dependencies (but that's a quite strong limitation which makes it suitable only for a minority of use cases).

File: applications/ISO20022-Implementation-POC.md [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+There are many ways in which this project can be extended in the future. Some of them are:
+
+### Protocol
+1. Allow for multiple currencies on the Polkadot Asset Hub
+2. Give customers direct access to their onchain funds
+3. Process a wider variety of ISO20022 transactions like card payments, currency conversions or even securities trading
+4. Make bank client PCI SSC compliant
+
+### Adoption
+1. Create partnerships with banks for pilot phase
```
Comment (takahser) [inline applications/ISO20022-Implementation-POC.md:238]: I see you're incorporated in Switzerland. Is that your target market? Do you have contacts in banks, and if yes, which ones would you target? If you're not confident sharing it here, you can also send us an email at grants AT web3 DOT foundation or schedule a call [here](https://calendar.app.google/puH98UkrfgEd66Ki9).

File: applications/ISO20022-Implementation-POC.md [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+
+### Project Details
+
+The goal of this project will be to implement a fully functional minimal viable product (MVP) that lets banks settle payments on the Polkadot blockchain using the ISO20022 standard. We will focus on customer-to-customer interbank payments and payment reversals, ie. Customer C1 with a bank account at bank B1 sending an amount of DOT to customer C2 with an account at bank B2 using ISO20022 messages.
+
+#### MVP Components
+This MVP will consist of the following components:
+
+##### 1. Substrate Pallet
+The first component is a pallet that will be resopnsible for storing hashed ISO20022 messages and handeling interbank payment settlements as well as a payment reversal mechanism.
```
Comment (takahser) [inline applications/ISO20022-Implementation-POC.md:25]:
```suggestion
The first component is a pallet that will be responsible for storing hashed ISO20022 messages and handling interbank payment settlements as well as a payment reversal mechanism.
```

File: applications/ISO20022-Implementation-POC.md [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+7. Client account management system
+    - Create for each client an account number
+    - Store for each client a list of hashes form the chain
+    - Functionality to derive client balance from onchain hashes
+8. Hashmap-like storage of all ISO messages and salts
+
+We will use NodeJs to build this MVP client.
+
+##### 3. Interbank Message Bus
+As the blockchain only stores hashes of the ISO 20022 messages in order to preserve privacy, we need some a  communication system to exchange ISO20022 messages between banks.
```
Comment (takahser) [inline applications/ISO20022-Implementation-POC.md:48]: So the blockchain only stores hashes of the pacs.008.001.11 messages, but the banks send each other the pacs.008.001.11 messages without hashing them peer-to-peer, did I understand correctly?

**Review by takahser — CHANGES_REQUESTED** [review]

File: applications/ISO20022-Implementation-POC.md (hunk 2 above)
Comment (pifragile) [reply applications/ISO20022-Implementation-POC.md:45]: I understand, yes I would be open for that.

File: applications/ISO20022-Implementation-POC.md (hunk 3 above)
Comment (pifragile) [reply applications/ISO20022-Implementation-POC.md:238]: Yes, swiss banks would be a natural place to start. Let's speak more in a call, just booked one.

File: applications/ISO20022-Implementation-POC.md (hunk 5 above)
Comment (pifragile) [reply applications/ISO20022-Implementation-POC.md:48]: Yes, that's correct.

**Review by semuelle — COMMENTED** [review]
> So I wanted to create a grant application for a project that is actually needed in the ecosystem and thus I replied to an RFP. I am not planning to use this client myself.

Thanks for the clarification. I really appreciate your interest in building this and in finding opportunities to improve the ecosystem. My concern is that many projects, especially ones targeting a niche, are often outdated before they are picked up again. We have funded many such projects, and therefore are now more thorough with evaluating the medium- and long-term prospects of projects. If someone were to, for example, pair this with six months of maintenance and outreach activities, it would make a great fit for the Decentralized Futures Program.

**Comment by semuelle** [comment]
Would you consider applying with the Decentralized Futures Program, @pifragile?

**Comment by pifragile** [comment]
> Would you consider applying with the Decentralized Futures Program, @pifragile?

Haha, I'm scared that if I say yes, you will reject this proposal😂 No, I mean sure, I could consider that.

**Comment by keeganquigley** [comment]
Hi @pifragile yes basically that is what would happen, but you could always re-apply if DF funding were not approved. There might be some benefits to that as well, if your aim is to eventually be a self-sustainable entity. Up to $50k is able to be fast-tracked to get it in front of the committee faster, and the DF program also has a broader scope, allowing you to fund additional areas such as audits, marketing, partnerships, etc. as a few examples.

**Comment by semuelle** [comment]
> Hi @pifragile yes basically that is what would happen, but you could always re-apply if DF funding were not approved.

We could even reopen this PR if necessary, so you wouldn't even have to submit a new application.

**Comment by pifragile** [comment]
@keeganquigley @semuelle Thanks for your clarifications! The way I understand the requirements for the DF program, I am not sure if my project in it's current form - which is intended to be a POC - is already mature enough for DF. Would be grateful to hear your thoughts on that.

**Comment by semuelle** [comment]
> Would be grateful to hear your thoughts on that.

We wouldn't have suggested it if we thought it wouldn't be a fit. :) "Kickstarting teams and initiatives" doesn't mean that you need to be self-sufficient at the end of the grant. It just means there should be potential and plans for beyond the grant.

**Comment by pifragile** [comment]
Ok, thanks @semuelle :) And thanks a lot to everyone for taking the time to consider my proposal @keeganquigley @takahser.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the effort you put into this application. As pointed out above, feel free to apply for DF funding. I'm closing this PR for now.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-iso-20022-integration/8483/1

**Review by semuelle — CHANGES_REQUESTED** [review]
Welcome back, @pifragile. To restart the discussion, I'll summarize the points that were raised so far:

- @takahser suggested an implementation in C# or Java. Any new insights/opinion on this?
- I asked about your concrete `Future Plans`. I also noticed that there is no mention of future funding. It would be interesting to hear how you are planning to fund further development.

I also have two more requests:
- As you said, most projects claiming to be ISO20022 compatible are pretty vague about how they achieve that. In that case, I think it would be interesting to spend some time researching what it _could_ mean and the differences in how they approach the topic. 
- Could you denominate the milestone costs in USD? We can pay in CHF, but it would be converted at the time of payment.

**Comment by semuelle** [comment]
pinging @pifragile

**Comment by pifragile** [comment]
@semuelle
- Java/C# is possible.
- We do not have any specific plans for this project. This was intended to be an early stage exploration of a complex topic.
- I agree that it would make sense to take a step back to figure out the best way to build this, but the time that I am willing to put into those proposals is quite saturated now.  Would you be willing to accept a 5k grant for a report on the current state of ISO2002 integrations of other blockchains and more concrete technical architecture/design proposals on different ways to achieve ISO compatibility?
- Sure, can do as soon as we come to an agreement on how to proceed best.

**Comment by semuelle** [comment]
> * Would you be willing to accept a 5k grant for a report on the current state of ISO2002 integrations of other blockchains and more concrete technical architecture/design proposals on different ways to achieve ISO compatibility?

Sounds very good to me. If you reduce the scope accordingly, I'll happily approve. It'll also only take two votes and we can revisit the implementation afterwards, if you're still up for it then!

**Comment by pifragile** [comment]
> > * Would you be willing to accept a 5k grant for a report on the current state of ISO2002 integrations of other blockchains and more concrete technical architecture/design proposals on different ways to achieve ISO compatibility?
> 
> Sounds very good to me. If you reduce the scope accordingly, I'll happily approve. It'll also only take two votes and we can revisit the implementation afterwards, if you're still up for it then!

Ok great. Should I use this proposal or open a new one?

**Comment by semuelle** [comment]
You can use this one, that's probably the easiest.

**Comment by pifragile** [comment]
> You can use this one, that's probably the easiest.

Ok @semuelle, done!

File: applications/ISO20022-Implementation-POC.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+### Milestone 1 — Ecosystem Research and Technical Design
+
+- **Estimated duration:** 1 month
+- **FTE:**  0.25 (5 person days)
+- **Costs:** 5200 CHF (5850 USD)
+
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| 0e. | Article | We will publish an **article** that outlines the current state of other blockchain's ISO20022 development as well as technical designs how to solve this for Polkadot.|
```
Comment (semuelle) [inline applications/ISO20022-Implementation-POC.md:150]: Thanks for the update, @pifragile. Happy to move forward. Just to give the article a bit of structure, I would like to ask that a bibliography and description of how the material was sourced be included. Can be part of the article.

```suggestion
| -----: | ----------- | ------------- |
| **0c.** | Methodology | Description of how the information was sourced. |
| **0e.** | Bibliography | List of sources |
| 1. | Article | We will publish an **article** that outlines the current state of other blockchain's ISO20022 development as well as technical designs how to solve this for Polkadot.|
```

Comment (pifragile) [reply applications/ISO20022-Implementation-POC.md]: sure, makes sense!

**Review by semuelle — APPROVED** [review]

**Comment by semuelle** [comment]
You should have just received an email from our KYB provider, @pifragile. Could you check?

**Comment by takahser** [comment]
Hi @pifragile, could you please fill out the [KYB form](https://in.sumsub.com/idensic/l/#/uni_JRqwo72Y5AjXaRDO))?

**Comment by pifragile** [comment]
> You should have just received an email from our KYB provider, @pifragile. Could you check?

Yep got it, and submitted again. I remember submitting this already last time, but somehow one document was not accepted and I couldn't figure out which document they actually needed (I have a LLC (GmbH) registered in Switzerland). I tried a couple of times with different documents but no success. Hope it will work this time :)

**Comment by semuelle** [comment]
Can confirm, check was successful.

**Pull request merged**